src/output/output.h

```diff
@@ -0,0 +1,162 @@
+
+// Value is an arbitrary value.
+//
+// This is a primitive, or a high-level aggregation over values. It is not
+// guaranteed to be orderable or comparable.
+struct Value {
+  // Histogram is a specific type of value, which contains a sequence
+  // of labels that represent the end interval for buckets, and a sequence of
+  // counts. All the labels are typically numbers, strings or `Symbolic`.
+  //
+  // The labels field represents the upper bounds for each of the buckets,
+  // and if counts is larger than labels then the top bucket has no upper bound.
+  //
+  // If there is a lower-bound for the first bucket, it is specified within the
+  // lower_bound field.
+  struct Histogram {
+    std::optional<Primitive> lower_bound;
+    std::vector<Primitive> labels;
+    std::vector<uint64_t> counts;
+  };
+
+  // OrderedMap defines a map that can be ordered externally, and therefore
+  // is not necessarily bound to be ordered by key.
+  struct OrderedMap {
+    std::vector<std::pair<Primitive, Value>> values;
+  };
+
+  // Stats is just an arbitrary classifier for another value. It changes
+  // only some of the encoding used by some outputs.
+  struct Stats {
+    Stats(OrderedMap&& m) : value(std::move(m)) {};
+    Stats(Primitive&& p) : value(std::move(p)) {};
+    std::variant<OrderedMap, Primitive> value;
+  };
+
+  // TimeSeries is a sequence of values that are collected over time.
+  struct TimeSeries {
+    std::vector<std::pair<Primitive::Timestamp, Primitive>> values;
+  };
+
+  using Variant = std::variant<Primitive,
+                               Histogram,
+                               std::vector<Value>,
+                               OrderedMap,
+                               Stats,
+                               TimeSeries>;
+
+  template <typename T>
+  Value(T&& t)
+    requires std::constructible_from<Variant, T>
+      : variant(std::forward<T>(t)){};
+
+  Variant variant;
+};
+
+// Abstract class for output.
+//
+// This should be overriden by individual implementations.
+class Output {
+public:
+  virtual ~Output() = default;
+
+  // Print high-level maps, with special support for histogram-type maps,
+  // and "stats" maps, where richer statistics are required. The exact
+  // statistics or method for publishing the histograms is up to the engine.
+  //
+  // The "stats" boolean indicates whether this should be considered as "stats"
+  // map or not, in order to preserve message types for JSON encoding.
+  virtual void map(const std::string& name, const Value& value) = 0;
+
+  // Print an arbitrary value.
+  virtual void value(const Value& value) = 0;
+
+  // Specialized messages during execution.
+  virtual void printf(const std::string& str) = 0;
+  virtual void time(const std::string& time) = 0;
+  virtual void cat(const std::string& cat) = 0;
+  virtual void join(const std::string& join) = 0;
+  virtual void syscall(const std::string& syscall) = 0;
```

Comment:
Is it possible to smush these all together into a method like `special` that takes an enum to account for these one-offs? Might be cleaner than these methods.

Reply:
I absolutely agree and hate the structure of this API. However, there's enough changing at once, so I'm not changing the "message" protocol here, and just doing a 1:1 mapping from the `MessageType` to virtual methods on the output side to preserve the exact behavior.

Reply:
I'm honestly not sure in which way I'd want to clean this up. For now, for better or worse, we have codified this series of messages in the JSON output. The methods here correspond 1:1 to the `type` emitted. I don't really want to break that `1:1` correspondence and have one special case for 9/10th of the messages. Either a set of virtual methods or a `enum` are fine, having a hybrid of **both** seems a bit silly. I just elected for the first one, since I think it is clearer this way (and yes, the kludginess of having a dedicated "printf" message type is apparent, but this is also by design).

What I *would* like to do is actually eliminate most of these messages, but I think that's bigger than a `TODO`.

Reply:
> that's bigger than a TODO

That's what issues are for 😉 but fine, I'm not blocking on this. Just another yak that needs shaving.